Add support for IAM token

vkuznet opened the pull request:
Fixes https://github.com/dmwm/WMCore/issues/10940

This PR provides support to include IAM token to HTTP requests. The token can be obtained via separate process, e.g. via curl, `create-iam-token.sh` script or [oidc-agent](https://indigo-iam.github.io/v/v1.7.0/docs/tasks/user/getting-a-token/), and then can be used by any HTTP client. We added appropriate `Authorization` HTTP header to `pycurl_manager`. The token itself can be provided either via file name or environment variable.

For more information please consult [Tokens in WMCore](https://github.com/dmwm/WMCore/wiki/Tokens-in-WMCore) wiki page.

#### Status
ready

#### Description
We provide a patch to pycurl manager to setup appropriate HTTP header if token is provided via configuration or environment. And, `TokenManager` module which provides different ways to obtain IAM token, either via `readToken` function or `TokenManager` class.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
Here is PR for wmcore/wmagent specs: https://github.com/cms-sw/cmsdist/pull/7647

#### External dependencies / deployment changes
None

vkuznet committed:
Add support for IAM token

vkuznet commented:
@amaltaro , @todor-ivanov  could you please review this PR. The changes are trivial and if merged we can start testing usage of WMCore services with token based authentication.

amaltaro commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -61,6 +61,14 @@
 from Utils.PortForward import portForward, PortForward
 
 
+def get_token(env=None):
```
Valentin, I believe at some point we will want to have different operations for tokens (like, renewing, checking validity, etc) and maybe even different technologies. Right?

In that case, I would suggest to create a new module under the `Utils` package and have all the token-based implementation in there.

amaltaro commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -242,6 +257,7 @@ def set_opts(self, curl, url, params, headers,
         if verbose:
             curl.setopt(pycurl.VERBOSE, True)
             curl.setopt(pycurl.DEBUGFUNCTION, self.debug)
+        self.request_headers = headers
```
Given that this attribute is not used anywhere but here in the `__init__` method, it looks like this line could be removed.

vkuznet replied:
the request headers can be required in different clients, e.g. DBSClient, and should be set accodringly. This is why it was set here. The are initially declared in ctor and set appropriate for every HTTP request.

amaltaro reviewed (changes requested):
Valentin, in addition to the comments made along the code, I have a more general question/comment on how these tokens will be used, like:
* once we load a token in the `RequestHandler` object, I think we should keep using it instead of making the os.environ call for every HTTP call that goes through pycurl_manager.
* with that said, we would need to have a mechanism to also update our token because it will eventually expire, right?
* adding support to reading this token from a file might be a good idea

Last but not least, the `pycurl_manager` module also contains some functions. We need to make sure `getdata` function is covered with this implementation.

In terms of other HTTP modules, we use this one for uploading a root file - from within the worker node - to the DQM Gui:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/HTTPS/HTTPSAuthHandler.py
you might want to have a look at it as well.

vkuznet commented:
@amaltaro thanks for comments. We're working on setting up never expired token, similar to how we handle proxy for x509. Once new service will be in place, the token will be updated on regular basis and placed in a file. The token validity is defined by IAM provider and as far as I saw it is valid for 1hour. Therefore, I'll adjust code to do the following:
- add token manager to provide valid token
   - it will read token from default env/file
   - it will check validity of the token
   - if necessary it will re-read token again (if token is expired)
- initialize token manager at ctor

I'll put new code in Utils as you suggested and then use it here and pycurl manager.

vkuznet commented:
@amaltaro , I provided a first draft of TokenManager which so far can only read and validate token. Later, we may discuss and add new functionality how to obtain new token in TokenManager. Please bare this in your review. I explored several python libraries and decided to stick to [pyjwt](https://github.com/jpadilla/pyjwt/) recommended on [jwt.io](https://jwt.io/libraries). This library provides ability to inspect token and does not have much dependencies. This implies that it should be added to requirements of WMCore.

vkuznet commented:
@amaltaro , @todor-ivanov I am not sure I understand the following pylint error `[E1601, line 72 in tokenData: print statement used]()`. According to quick Google search it is false positive, see [here](https://github.com/PyCQA/pylint/issues/437)

amaltaro commented:
@vkuznet I haven't looked into this PR, but please replace the print statement by the logging library. This is an effort that we made a few years ago and kept the print statement only where we didn't manage to get a logger object working well.

amaltaro commented:
@vkuznet Valentin, could you please work on the cmsdist specs to get this new python library in the stack? Once you have that working, please link it to the first description of this PR. I haven't looked into these details, but I believe the massive failures are coming from the import error for the new library.

vkuznet commented:
@amaltaro I made necessary changes to replace print with logger. I checked that all pylint issues are gone. But the jenkins tests are failing due to the following:
```
...
  File "/build/cmsbld/jenkins/workspace/DMWM-WMCorePy3-PR-unittests/SLICE/0/label/cms-dmwm-cc7/code/src/python/WMCore/Services/Requests.py", line 80, in __init__
    self.reqmgr = RequestHandler()
name 'RequestHandler' is not defined
```
It seems to me that in Jenkins there is an issue with environment since `RequestHandler` is defined in `WMCore/Services/Requests.py` as following:
```
try:
    from WMCore.Services.pycurl_manager import RequestHandler, ResponseHeader
except ImportError:
    pass
```
Therefore, it seems to me that there is an issue with loading pycurl_manager which is part of this PR. Since there is not any printout in import error section we don't know what is the actual problem. Locally, I can easily load pycurl_manager and its pylint scores 10/10. If you want I can change `Requests.py` to add import error printout to understand the issue.

vkuznet commented:
@amaltaro , please be specific which spec file you're talking about? I adjusted code base to handle missing `jwt` import, hopefully this will fix failed unit tests. I think we should add separate PR for requirements.txt and the spec, but I need to know which spec(s) do you use these days.

vkuznet commented:
Here is new PR for specs: https://github.com/cms-sw/cmsdist/pull/7647

vkuznet commented:
Now, majority of unit tests failures are gone. The rest 5 failed tests are not related to this PR since they report issues with workflows which have nothing to do with proposed changes. I suggest that @amaltaro should look at those failure to decide what it is.

vkuznet commented:
@amaltaro I rebased this PR against master and added `pyjwt` dependency here too. Please review again.

amaltaro reviewed (changes requested):
Valentin, I have a few comments along the code.

In addition to that, I have some more general comments/questions:
* can you please improve those docstrings? This is going to be a common and shared library, so we better make it clear to anyone that starts looking into these modules.
* second, I wonder why you created those functions in the TokenManager module? Don't you think it's better to make it methods of the TokenManager class?
* you have encapsulated a few calls in try/except blocks. I believe it's only for the tests to pass, right? If that's the case, we could merge the cmsdist changes such that we get this library in our stack, and run these tests again (without try/except).

I will likely have further feedback for you, but it's probably better to work these out first. Thanks

vkuznet commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,132 @@
+        """
+        Token manager reads IAM tokens either from file or env.
+        It caches token along with expiration timestamp.
+        By default the env variable to use is IAM_TOKEN.
```
done

vkuznet commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,132 @@
+        self.expire = 0
+        self.token = None
+        self.logger = logger if logger else logging.getLogger()
+        try:
```
we don't have yet mechanism how to create token as code only reads existing one. Once we'll add code to create tokens then we'll add real failure procedure

vkuznet commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -213,6 +219,14 @@ def set_opts(self, curl, url, params, headers,
             raise Exception('Unsupported HTTP method "%s"' % verb)
 
+        try:
+            token = self.tmgr.getToken()
```
It really depends how we decide to handle token creation. It is not the same as X509. Current procedure is to have separate process/daemon which will handle token creation. it is more like ssh-agent rather the voms-proxy-init. In former case, you deploy separate token agent which handle the tokens. In latter, you really obtain a new proxy.

amaltaro replied:
Yes, what I am saying is that if the system is defined to use tokens, it should use it instead of the x509 certificates, even if that token is expired!

So, I would basically take this code out of the try/except block.

amaltaro replied:
Resolving this comment as you have already implemented the suggested changes, thanks.

vkuznet commented:
@amaltaro to answer your questions:
- I added/improved docstrings
- the functions are separate from TokenManager class since they are only need to read tokens. Depending how you'll manage tokens the TokenManager may create new tokens or refresh it, etc. Therefore, someone who will use this library can use either function, e.g. only import function to read token from env, or have TokenMamanger to handle token entirely.
- try/except added if infrastructure does not provide any token and code should function as is. I think we still need it until we'll setup infrastructural pieces in place
- token may or may not be presented in HTTP request, e.g. we still have period when we'll need to work with x509, therefore we still need try/except block regardless of tokens during the transition period
- token management is different from x509, as I wrote IAM provide token daemon similar to ssh agent, and it will handles the tokens for you. We only need to read token which will be managed by external software. Said that, we may also need to add functionality to obtain tokens in env which will not have token manager, that's why we may need and extend TokenManager.
- I removed requests_headers and let jenkins complain if it finds missing dependencies. The DBSClient relies on RestClient which by itself uses pycurl manager and headers are dumped in HTTPError.

amaltaro commented:
Valentin, side question, in addition to the `oidc-agent` service that we can install on these nodes to provide us with tokens. Would there be another option using the current software stack (or with this pyjwt library)?

vkuznet commented:
Tokens can be obtained in a different way and from different providers. For instance, the auth proxy server (aps) I'm pushing to replace our apache stack is capable to work with tokens from CERN SSO and IAM. As such, there are different ways to get tokens. For instance, someone can get CERN SSO token via web interface, e.g.  on aps server someone can visit `/token` end-point and you'll get your token. If you're using IAM, you can use either oidc-token, or you may request client credentials to get token from IAM directly (e.g. we created token manager pod for k8s which is doing exactly that and it will keep fresh token for all services in k8s infrastructure). I think using client credentials (which is client id and client secret) it is possible to use 3rd party library, e.g. pyjwt, to obtain token, but it will require coding and management on your infrastructure (someone will need to securely handle client credentials).

But I think ultimately we'll rely on oidc agent to provide us tokens to reduce overhead of management them. In k8s, we already have token manager for that which works with IAM. I also provide token manager as part of aps codebase (but it is not python it is written in Go), and it works with aps+CERN SSO.

amaltaro commented:
test this please

amaltaro reviewed (changes requested):
This is not a full review, but could you please look into these failing unit tests?

I had a very quick look at the `oidc-agent` service as well and the installation procedure, and I do not see any RPM installation instructions. Can't this complicate deployment of this in a non-containerized environment? 

WMAgents can run in different OS distributions as well - FNAL vs CERN nodes - and this is another item to be taken into consideration.

Having said that, unless it requires a substantial amount of effort, I would say that implementing our own code to fetch a token would be the best (we should maintain one credential for CERN agents, one for FNAL agents). This is more or less what has been done for Rucio here:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/Tools/PycurlRucio.py#L65-L108

vkuznet commented:
regarding rpm based install for token tools, can it be as simple as `yum search oidc` and then `yum install oidc...`? Regarding Rucio token, it is not the same as OIDC one (as far as I know), therefore you will still need to manage tokens. Said that, on k8s we already do that and everything is ready for services to use it. For WMAgent we can provide a daemon too, but I think you just need to try first yum approach. Please read [install instructions](https://indigo-dc.gitbook.io/oidc-agent/installation/install) which basically confirms what I just wrote:
```
CentOS 7
sudo yum install \
      libcurl-devel \
      libsodium-devel \
      libsodium-static \
      libmicrohttpd-devel \
      libseccomp-devel \
      libsecret-devel \
      libqrencode-devel
```

vkuznet commented:
test this please

vkuznet commented:
@amaltaro , I looked at failed tests and they are related to Unauthorized CouchDB access, see https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/12936/testReport/junit/WMComponent_t.ErrorHandler_t.ErrorHandler_t/ErrorHandlerTest/testD_Exhausted/
Did you change anything for couch access in jenkins. I have no idea why proposed changes can affect couch. The only change is to pycurl_manager which is within tr/except block which pass if token is not present. Therefore it can't affect anything. The `TokenManager` class is independent from the existing WMCore and can't affect either. Therefore, I don't see any logical reason how the proposed code affects the CouchDB access. Any clue?

amaltaro commented:
The only change that went in in the last week'ish was this:
https://github.com/dmwm/WMCore/pull/11045

and it did not touch ErrorHandler code. Are you able to reproduce this error when testing with WMCore container unit test setup? Or it only shows up in jenkins?

vkuznet commented:
Alan, do you have any documentation how to setup "WMCore container unit test setup"? I never did it and have no idea how to do it. Moreover, you should also explain how to do it with and without #11045 PR.

vkuznet commented:
Alan, after lots of digging and following a different documentation [wiki page](https://github.com/dmwm/WMCore/wiki/setup-wmcore-unittest#1-using-the-wmagent-docker-container) I setup WMCore container where I executed failed test, and it is failing like this:

```
# I patched WMCore with changes I provide in this PR

# I started WMCore container

# I run one of the tests 
python3 /home/dmwm/wmcore_unittest/WMCore/test/python/WMComponent_t/ErrorHandler_t/ErrorHandler_t.py
....
http.client.HTTPException: url=http://localhost:6994/job_summary/_design/WMStatsAgent/_update/jobSummaryState/7b07d232-b161-4419-9507-54a42e2f6b70?newstate=jobcooloff&timestamp=1648584707, code=404, reason=Object Not Found, headers={'Server': 'CouchDB/1.6.1 (Erlang OTP/R16B03-1)', 'Date': 'Tue, 29 Mar 2022 20:11:47 GMT', 'Content-Type': 'application/json', 'Content-Length': '41', 'Cache-Control': 'must-revalidate'}, result=b'{"error":"not_found","reason":"missing"}\n'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/JobStateMachine/ChangeState.py", line 178, in propagate
    self.recordInCouch(jobs, newstate, oldstate, updatesummary)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/JobStateMachine/ChangeState.py", line 304, in recordInCouch
    self.jsumdatabase.makeRequest(uri=updateUri, type="PUT", decode=False)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Database/CMSCouch.py", line 135, in makeRequest
    self.checkForCouchError(getattr(e, "status", None),
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Database/CMSCouch.py", line 153, in checkForCouchError
    raise CouchNotFoundError(reason, data, result)
WMCore.Database.CMSCouch.CouchNotFoundError: CouchNotFoundError - reason: Object Not Found, data: None result: None
```

The output seems identical to what jenkins provide, therefore I would conclude that something is related to CouchDB setup in a container. Please advise how to proceed forward with it.

vkuznet commented:
Alan, please respond to this ticket on how to move forward with tests. I provided output from WMCore container as well.

vkuznet commented:
@amaltaro any suggestions how to move forward with it? I'm awaiting your feedback on how to proceed with this PR.

amaltaro commented:
test this please

amaltaro commented:
Out of the 5 new failing unit tests, only this:
```
WMComponent_t.TaskArchiver_t.TaskArchiver_t.TaskArchiverTest:testA_BasicFunctionTest
```

fails when running from the docker image. The other 4 tests are successful, with or without your patch.
Jenkins is complaining about:
```
CouchUnauthorisedError - reason: Unauthorized, data: {} result: None
```

I don't think these failures are coming from this PR, but it needs further investigation. One thing I can say, our unit tests aren't behaving well for a few weeks now (we keep seeing the "tests added", even though it's been weeks that those have been added).

vkuznet commented:
ok, from the error message it seems it is related to Couch access and most likely expired certificate in jenkins. But since tests do not tell which Couch it contacts I can't verify that. If you know the CouchDB address we may check FE logs.

amaltaro commented:
Besides a some services/calls that are not mocked, our unit tests are pretty much self container. Meaning, we deploy MariaDB and CouchDB at the localhost, run the unit tests, and clean up everything.

I'm adding a debugging statement in the jenkins setup, let me trigger tests once again.

amaltaro commented:
test this please

amaltaro commented:
For the record, this log has a few CouchUnauthorized errors:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCorePy3-PR-unittests/SLICE=9,label=cms-dmwm-cc7/lastBuild/consoleText

@vkuznet can you please add a print statement above this line:
https://github.com/dmwm/WMCore/blob/eba0a315ed973616357e231976f7092adcb6b2e6/src/python/WMCore/JobStateMachine/ChangeState.py#L92

printing the value of `self.config.JobStateMachine.couchurl`? Push this in as a new commit please.

vkuznet commented:
test this please

vkuznet commented:
Alan, tests are now passing. There are two warnings about unit tests but I don't know if you want to fix them. I put explicitly few comments about the code and we need to decide how to run unit tests when we'll handle real tokens. Feel free to make suggestions.

amaltaro commented:
@vkuznet Valentin, I am going to review it once again tomorrow. For now, could you please squash your commits (actually when rebasing it, you can simply delete the last 2 commits and force push it). But please do whatever GH workflow works for you.

vkuznet commented:
For completeness here are details about new token manager services on k8s:
- we created a new token manager services which is based on a regular cron job
- it creates new token on periodic interval and update a given file (which hold the token) on k8s cluster
- the new service is deployed on all cmsweb clusters
All details can be found in this ticket: https://its.cern.ch/jira/browse/CMSKUBERNETES-129

If DMWM wants to adapt this strategy outside of k8s infrastructure, e.g. on WMAgent nodes, then the following set of actions are required:
- install `oidc` token based tools based on set of instructions from IAM, see [here](https://indigo-iam.github.io/v/v1.7.0/docs/tasks/user/getting-a-token/). The RPMs or other installation methods are available.
- register your services or application with IAM provider and obtain its credentials (client_id and secret)
- run the following command on a node to obtain token on regular interval
```
# adjust client_id and client_secret accordingly to your setup
export client_id=`cat /etc/client/client_id`
export client_secret=`cat /etc/client/client_secret`

# obtain new token using client credentials
curl -s -d grant_type=client_credentials -d scope="profile" -u ${client_id}:${client_secret} https://cms-auth.web.cern.ch/token | jq -r '.access_token' > /tmp/token
```

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,149 @@
+    Get IAM token from environment
+    :param env: environment name to read the token, if not provided we'll use IAM_TOKEN
+    """
+    if env:
```
It can be refactored to:
```
    if env:
        return os.environ.get(env)
    return os.environ.get("IAM_TOKEN")
```

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,149 @@
+        return os.environ.get("IAM_TOKEN")
+    return None
+
+def readTokenFile(fname=None):
```
`fname` should be camel case --> `fName`. Please rework the docstring as well.

amaltaro reviewed (changes requested):
@vkuznet Valentin, this is another partial review because I think there are a few things worth changing first. 

I made some comments along the code, but regarding those functions in the TokenManager module, the `getToken` function is so small that I think we should  merge `readTokenEnv` and `readTokenFile`. Which will also minimize code duplication.

Given that TokenManager is a new code, please run an code reformat on it such that it comes out as clean as possible. For instance, spacing between functions are meant to be 2 blank lines.

Regarding pycurl_manager, I am in favor of: if the client specifies that it wants to use tokens, then requests should either be made with it, or fail. From the code implemented, I think it would fallback to the certificates, no?

vkuznet commented:
@amaltaro I addressed all issues you raised except merging `readTokenEnv` and `readTokenFile` functions. I think they should be kept separately since there may be different use-cases for them. We still don't know mode of operations we'll end-up with and both `env` and `file` based ones are possible. As such, I still think we need both functions to use for different purposes.

amaltaro commented:
> @amaltaro I addressed all issues you raised except merging readTokenEnv and readTokenFile functions. I think they should be kept separately since there may be different use-cases for them. We still don't know mode of operations we'll end-up with and both env and file based ones are possible. As such, I still think we need both functions to use for different purposes.

Thanks Valentin. And I agree with that! Sorry for not being clear, what I meant was to merge those two functions and move their code under the `getToken` function, which already has some of that logic. This will make the code more clear and less redundant IMO.

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,130 @@
+        key,
+        algorithms=[alg],
+        audience=aud,
+        options={"verify_exp": True},
```
can you please add a comment on what this `verify_exp` means and is used for?

vkuznet replied:
it is parameter of jwt linbrary to check token validity

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,130 @@
+    )
+    return data
+
+def validToken(token):
```
Perhaps calling it `isTokenValid` would be better?

Somehow unrelated to this, but we should also have a function/method that returns the lifetime of a given token (such that we can get a new one *before* we see it expired)

vkuznet replied:
ok

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,130 @@
+def validToken(token):
+    """
+    validate token expiration timestamp
+    :param token: token string
```
docstring needs some rework.

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,130 @@
+        return token
+    if name:
+        return os.environ.get(name)
+    return os.environ.get("IAM_TOKEN")
```
I'd suggest to actually suggest a global constant with this default environment variable name, e.g. (at the top of the module):
```
DEFAULT_TOKEN_ENV="IAM_TOKEN"
```
what do you think?

vkuznet replied:
to Alan: why do you want to invent and introduce yet another global variable, I always against all globals.
to Todor: why do you want to complicate handling of different use-case by adding more code? In current implementation you can either read token from env or from file using the same function. In your proposal, not only we need two separate functions here, but also on a client side you'll need to call them independently. Why do you want to duplicate behavior while gaining nothing?

amaltaro replied:
Thanks Valentin, I'm resolving this as well.

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,130 @@
+        It caches token along with expiration timestamp.
+        By default the env variable to use is IAM_TOKEN.
+        :param name: string representing either file or env where we should read token from
+        :param url: IAM provider URL, by default it is https://cms-auth.web.cern.ch/jwk
```
Same comment as the one above, for these default urls.

amaltaro replied:
Another reminder for the default urls (similar to the one provided in the function above).

amaltaro replied:
Regarding this comment, please actually consider the one made in the `__init__` method signature.

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,130 @@
+        except Exception as exc:
+            self.logger.exception(str(exc))
+            raise
+        self.expire = tokenDict.get('exp', 0)
```
Along the comment made above - for creating a new function - now that I look into this, it makes sense to me to have a method called `getExpirationTime` or `getLifetime` to read this attribute.

vkuznet replied:
no, it should be `isValidToken` since we read token and it can be mailformed, and only if we read it correctly we verify its expiration timestamp. So, let's not change it.

amaltaro replied:
Yes, what I wanted to say is that we should have something like:
```
def getLifetime(self):
    return self.expire - int(time.time()) 
```

The use case here is that we want to renew this token BEFORE it gets expired. So we need to have a mechanism to constantly poll how long it's still valid for.

amaltaro replied:
Resolving this as well, since you implemented this method below. Thanks

amaltaro commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -165,6 +166,12 @@ def __init__(self, config=None, logger=None):
         self.followlocation = config.get('followlocation', defaultOpts['FOLLOWLOCATION'])
         self.maxredirs = config.get('maxredirs', defaultOpts['MAXREDIRS'])
         self.logger = logger if logger else logging.getLogger()
+        self.request_headers = None
```
This seems not to be necessary. You might want to remove the TMP commented out code as well (from 100 lines below).

vkuznet replied:
done

amaltaro reviewed (changes requested):
Valentin, thanks for making those changes.

I have further comments for your consideration.

In addition to those, I have a few more general comments:
1) is there any reason to define both functions and class/methods for the TokenManager? Do you see any problems in actually moving those functions as a TokenManager method? IMO, the code looks more organized and clear when we have methods defined in a class; and we also can profit from the logger object and have proper logging in place.
2) please run a code reformat in the new module (TokenManager.py). pep8 is failing to run in jenkins, but I see a few places that we could have enhance and avoid suggestions from jenkins (like double spaces between modules). Together with this, unless there is really no better option, I'd say that variables with 3 chars only are usually not the most meaningful and cleaner option.
3) we can break things down in smaller parts, but I think that having the capability of creating a token without any external service/daemon would be extremely useful (and reduce dependencies on our stack/jenkins/dev setup).

todor-ivanov reviewed (commented):
Hi @vkuznet for providing all this construct. This functionality is really needed. I have left a comment or two inline, but in general they all go in the same direction, which is: 
I think having the IAM token provided through the environment would be the best approach. And from the two possible options:
* One - the whole token stored in an environment variable
* The other - the whole token stored in a file and pointed to by an environment variable holding the filename
I'd say I am more in favor of the later. But having them both implemented and having the flexibility to chose of course would be the best. 

Thanks!

vkuznet commented on requirements.txt:
```diff
@@ -41,3 +42,4 @@ retry~=0.9.2                  # wmcore,wmagent,wmagent-devtools,reqmgr2,reqmon,g
 rucio-clients~=1.25.5         # wmcore,wmagent,global-workqueue,reqmgr2ms
 Sphinx~=4.0.3                 # wmcore,wmagent,wmagent-devtools,reqmgr2,reqmon,acdcserver,global-workqueue,reqmgr2ms
 SQLAlchemy~=1.3.3             # wmcore,wmagent
+pyjwt~=2.3.0                  # wmagent,wmagent-devtools,reqmgr2,reqmon,acdcserver,global-workqueue,reqmgr2ms
```
ok

vkuznet commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,130 @@
+        return os.environ.get(name)
+    return os.environ.get("IAM_TOKEN")
+
+def tokenData(token, url=None, aud=None):
```
fine

vkuznet commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,130 @@
+        return {}
+    if isinstance(token, str):
+        token = bytes(token, 'utf-8')
+    if not url: # use CMS IAM
```
yes

vkuznet commented:
I adjusted code based on your comments, but you also asked several questions which I answer below:
Alan's questions:
- class vs functions, the `TokenManager` class will handle token. We said since we don't know yet which behavior we want to token manager we keep it as a class. On a contrary, if client wants to read a token it does not need entire class and can use simple function. Once we understand behavior we want we can adjust to one or another. But, since functions does not require any class member data it is simpler and less costly to keep it as simple function.
- pep8 autoformat is applied now
- if we want to manage tokens within python code then you'll need:
   - add code to obtain token (this can be done via curl call) but for that you'll need to manage client's credentials
   - you'll need to extend `TokenManager` to do this job (which so far is done by external tool)
   - you'll need to maintain TokenManager objects/instances within all different applications which will require token instead of using token oidc tools
   - bottom line, if you want independence from external tool you'll basically need to write its functionality and maintain it. I doubt it is productive. We got tools exactly for that, e.g. you don't manage you ssh keys, you don't need to manage tokens, just use existing tools

Todor's questions:
- I doubt that support of both use-cases (env and file) via separate code base will make code easier, you'll need to implement both use-case within token manager module and on a client side, you'll need to decide which method should come first, which second, what to do if one fail, etc. By unifying both your client does not care how token will be provided, i.e. either in env or in file. Moreover, file can be updated by external daemon, we may have different use-cases, e.g. in k8s we'll provide you a file which will be updated so you don't need to manage it, while on WMA you may decide to use env and obtain token in a different. way. Bottom line, so far you did not provide any benefits that separation will be easier to maintain.

vkuznet commented:
@amaltaro , @todor-ivanov , I replied to your comments. Please review this PR again and let's move it forward.

todor-ivanov commented:
Thanks @vkuznet! 
I did try to take a look at the latest changes of yours, but it seems a huge merge is obscuring the picture: https://github.com/dmwm/WMCore/pull/10939/commits/d99704fa846463daeb361791ecbadf6cce72a316 
Is this due to a merge of `master` branch  on top of a  `development` branch or due to a rebase? This may also affect the master branch history as well.

vkuznet commented:
@todor-ivanov , at some point I rebased it to master and this is probably what you see, but this PR does not show any of that, i.e. I can only see my changes, `Files changed 4` which reflects only my changes. Can you be more specific about [d99704f](https://github.com/dmwm/WMCore/pull/10939/commits/d99704fa846463daeb361791ecbadf6cce72a316) as it is not shown in `Files changed` on github.

todor-ivanov commented:
Hi @vkuznet sorry, .. indeed it is not shown at the 'Files changed` tab. I was looking at the link from the github mail I received  and it was pointing to the full rebase diff .... which was terrifying at a first glance. I am looking at the new changes now.

todor-ivanov reviewed (approved):
Looks good to me.
Thanks @vkuznet

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,135 @@
+    except:
+        traceback.print_exc()
+        raise
+    exp = tokenDict.get('exp', 0)
```
Can you please add a comment on what this `exp` attribute is supposed to be?
I guess it stands for `expires`, with an integer EPOCH time of when it expires?

vkuznet replied:
done

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,135 @@
+        try:
+            self.token = self.getToken()
+        except Exception as exc:
+            self.logger.exception(str(exc))
```
Can you add some more context to this error message? Something like:
```
self.logger.exception("Failed to get token. Details: %s", str(exc))
```
?

vkuznet replied:
done

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,135 @@
+ssl._create_default_https_context = ssl._create_unverified_context
+
+
+def getToken(name=None):
```
Now that I am getting to the end of this review. I see that we have this `getToken` function but the class below also has a `getToken` method. Even though we need a `TokenManager` object (self) to call the `getToken` method, I think it would be less confusing to rename this function to actually `readToken`.
What do you think?

vkuznet replied:
done

amaltaro reviewed (changes requested):
Valentin, please find further comments along the code.
I also wanted to mention that we have integrated `pyjwt` into our jenkins setup, in case you want to try more realistic tests.

vkuznet commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,135 @@
+    """
+    if name and os.path.exists(name):
+        token = None
+        with open(name) as istream:
```
done

vkuznet commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,135 @@
+    return os.environ.get("IAM_TOKEN")
+
+
+def tokenData(token, url=None, audUrl=None):
```
done

vkuznet commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,135 @@
+    """
+    if not token or not jwt:
+        return {}
+    if isinstance(token, str):
```
done

amaltaro replied:
Valentin, you can remove the `isinstance` check in the line above because it's already performed within the `encodeUnicodeToBytes` function.

amaltaro commented on src/python/Utils/TokenManager.py:
```diff
@@ -0,0 +1,139 @@
+    TokenManager class handles IAM tokens
+    """
+
+    def __init__(self, name=None, url=None, audUrl=None, logger=None):
```
```suggestion
    def __init__(self, name=None, url="https://cms-auth.web.cern.ch/jwk",
                        audUrl="https://wlcg.cern.ch/jwt/v1/any", logger=None):
```

this is what I am suggesting in my comment below (maybe the spacing in my comment is not correct, so please consider it as well).

vkuznet replied:
done

amaltaro reviewed (changes requested):
Valentin, I had another look at this and it's looking good now, almost ready to go.
I resolved a few comments that you have already worked on but were not automatically closed by GH, and left 2 or 3 comments that are worth it.

Once you have those changes in place, feel free to rebase and squash these commits. If it's not too late, please keep the test changes in a separate commit.

amaltaro commented on requirements.txt:
```diff
@@ -41,3 +41,4 @@ retry~=0.9.2                  # wmcore,wmagent,wmagent-devtools,reqmgr2,reqmon,g
 rucio-clients~=1.25.5         # wmcore,wmagent,global-workqueue,reqmgr2ms
 Sphinx~=4.0.3                 # wmcore,wmagent,wmagent-devtools,reqmgr2,reqmon,acdcserver,global-workqueue,reqmgr2ms
 SQLAlchemy~=1.3.3             # wmcore,wmagent
+pyjwt~=2.3.0                  # wmagent,wmagent-devtools,reqmgr2,reqmon,acdcserver,global-workqueue,reqmgr2ms,wmcore
```
Luckily I just looked this up in Pypi and noticed that this name isn't correctly written. As far as I know, Pypi is case sensitive, so please update it to `PyJWT`, from: https://pypi.org/project/PyJWT/

vkuznet replied:
done

amaltaro commented:
@vkuznet Valentin, can you also please give a brief summary of what is provided in this PR and what kind of token management it allows now? It'd be useful to link the WMCore token wiki as well: https://github.com/dmwm/WMCore/wiki/Tokens-in-WMCore

FYI @belforte @mapellidario this change is likely going in in the next release and it depends on a new python 3rd-party library called: `PyJWT`, which has been already integrated to `cmsdist` repo, but our wmcore/crab specs still need to have this dependency added to the Requires list.

Last but not least, these changes - especially to pycurl_manager - should be backwards compatible, so even if we are not using tokens and don't have `PyJWT` library, nothing should break. Valentin can confirm this though.

vkuznet commented:
@amaltaro , thanks for suggestions I adjusted the codebase accordingly. Regarding PR description. I updated it too. But this PR does not provide any token management, we only provide methods (either via function or class) to read token, but we do not manage tokens. Pointer to wiki has been added.

amaltaro commented:
Thank you, Valentin. Yes, by management I meant that we have some utilities to load in a token and check expiration time. But I agree this is only a small set of managing tokens.

If you are ready with these changes, please:
1) squash the commits accordingly
2) remove the no longer relevant GH labels
3) update the status `In development` to `ready` or `done`. 
4) and fire a final request review.

vkuznet commented:
test this please

amaltaro reviewed (approved):
Thanks, Valentin.

belforte commented:
thanks for keeping us in the loop. Appreciated.
